serde/compat.py

```diff
@@ -385,8 +385,14 @@ def recursive(cls: TypeLike) -> None:
             lst.add(Union)
             for arg in type_args(cls):
                 recursive(arg)
-        elif is_list(cls) or is_set(cls):
+        elif is_list(cls):
+            lst.add(List)
+            args = type_args(cls)
+            if args:
+                recursive(args[0])
+        elif is_set(cls):
             lst.add(List)
+            lst.add(Set)
```

Comment:
Do we need `lst.add(List)` here? :thinking:

Reply:
I kept it there as it was the original behaviour to have `List`. I cannot see it used in the generated code and it works without it. If you think it is safe to add only `Set` I will remove it.

Ran the test suites without it and all passed

Reply:
For reference, here is a minimal example and its generated code
```
from serde import SerdeSkip, serde, from_dict, Coerce


def custom(cls, o):
    raise SerdeSkip()


@serde(type_check=Coerce)
class Foo:
    bar: set[str]
```
```
==================================================
                       Foo
==================================================

--------------------------------------------------
          Functions generated by pyserde
--------------------------------------------------
def to_iter(obj, reuse_instances=None, convert_sets=None):
    if reuse_instances is None:
        reuse_instances = True
    if convert_sets is None:
        convert_sets = False
    if not is_dataclass(obj):
        return copy.deepcopy(obj)

    return (
        [coerce(str, v) for v in obj.bar] if convert_sets else set(coerce(str, v) for v in obj.bar),
    )

def to_dict(obj, reuse_instances=None, convert_sets=None):
    if reuse_instances is None:
        reuse_instances = True
    if convert_sets is None:
        convert_sets = False
    if not is_dataclass(obj):
        return copy.deepcopy(obj)

    res = {}
    res["bar"] = (
        [coerce(str, v) for v in obj.bar] if convert_sets else set(coerce(str, v) for v in obj.bar)
    )

    return res

def typecheck(self):
    if not is_instance(self.bar, typing.Set[str]):
        raise SerdeError(f"Foo.bar is not instance of typing.Set[str]")

    return

def literal_de_(
    cls=cls,
    maybe_generic=None,
    maybe_generic_type_vars=None,
    data=None,
    variable_type_args=None,
    reuse_instances=True,
):
    if data in ():
        return data
    raise SerdeError("Can not deserialize " + repr(data) + " as Literal[].")

def from_iter(
    cls=cls,
    maybe_generic=None,
    maybe_generic_type_vars=None,
    data=None,
    variable_type_args=None,
    reuse_instances=None,
):
    if reuse_instances is None:
        reuse_instances = True

    maybe_generic_type_vars = maybe_generic_type_vars or None

    __bar = set(coerce(str, v) for v in data[0])

    try:
        return cls(
            __bar,
        )
    except Exception as e:
        raise UserError(e)

def from_dict(
    cls=cls,
    maybe_generic=None,
    maybe_generic_type_vars=None,
    data=None,
    variable_type_args=None,
    reuse_instances=None,
):
    if reuse_instances is None:
        reuse_instances = True

    maybe_generic_type_vars = maybe_generic_type_vars or None

    __bar = set(coerce(str, v) for v in data["bar"])

    try:
        rv = cls(
            __bar,
        )
    except Exception as e:
        raise UserError(e)

    return rv


--------------------------------------------------
           Function references in scope
--------------------------------------------------
to_iter: <function to_iter at 0x103fe3e50>
to_dict: <function to_dict at 0x1043c3d30>
typecheck: <function typecheck at 0x103fdbee0>
literal_de_: <function literal_de_ at 0x1043dc670>
from_iter: <function from_iter at 0x103faf430>
from_dict: <function from_dict at 0x1043b10d0>

```

Reply:
For reference, here is the minimal example and the generated code
```
from serde import SerdeSkip, serde, from_dict


def custom(cls, o):
    raise SerdeSkip()


@serde(serializer=custom, deserializer=custom)
class Foo:
    bar: set[str]
```
```
==================================================
                       Foo
==================================================

--------------------------------------------------
          Functions generated by pyserde
--------------------------------------------------
def to_iter(obj, reuse_instances=None, convert_sets=None):
    if reuse_instances is None:
        reuse_instances = True
    if convert_sets is None:
        convert_sets = False
    if not is_dataclass(obj):
        return copy.deepcopy(obj)

    return (
        serde_custom_class_serializer(
            Set[str],
            obj.bar,
            default=lambda: [
                serde_custom_class_serializer(str, v, default=lambda: v) for v in obj.bar
            ]
            if convert_sets
            else set(serde_custom_class_serializer(str, v, default=lambda: v) for v in obj.bar),
        ),
    )

def to_dict(obj, reuse_instances=None, convert_sets=None):
    if reuse_instances is None:
        reuse_instances = True
    if convert_sets is None:
        convert_sets = False
    if not is_dataclass(obj):
        return copy.deepcopy(obj)

    res = {}
    res["bar"] = serde_custom_class_serializer(
        Set[str],
        obj.bar,
        default=lambda: [serde_custom_class_serializer(str, v, default=lambda: v) for v in obj.bar]
        if convert_sets
        else set(serde_custom_class_serializer(str, v, default=lambda: v) for v in obj.bar),
    )

    return res

def typecheck(self):
    if not is_instance(self.bar, typing.Set[str]):
        raise SerdeError(f"Foo.bar is not instance of typing.Set[str]")

    return

def literal_de_(
    cls=cls,
    maybe_generic=None,
    maybe_generic_type_vars=None,
    data=None,
    variable_type_args=None,
    reuse_instances=True,
):
    if data in ():
        return data
    raise SerdeError("Can not deserialize " + repr(data) + " as Literal[].")

def from_iter(
    cls=cls,
    maybe_generic=None,
    maybe_generic_type_vars=None,
    data=None,
    variable_type_args=None,
    reuse_instances=None,
):
    if reuse_instances is None:
        reuse_instances = True

    maybe_generic_type_vars = maybe_generic_type_vars or None

    __bar = serde_custom_class_deserializer(
        Set[str], data, data[0], default=lambda: set(v for v in data[0])
    )

    try:
        return cls(
            __bar,
        )
    except Exception as e:
        raise UserError(e)

def from_dict(
    cls=cls,
    maybe_generic=None,
    maybe_generic_type_vars=None,
    data=None,
    variable_type_args=None,
    reuse_instances=None,
):
    if reuse_instances is None:
        reuse_instances = True

    maybe_generic_type_vars = maybe_generic_type_vars or None

    __bar = serde_custom_class_deserializer(
        Set[str], data, data.get("bar"), default=lambda: set(v for v in data["bar"])
    )

    try:
        rv = cls(
            __bar,
        )
    except Exception as e:
        raise UserError(e)

    return rv


--------------------------------------------------
           Function references in scope
--------------------------------------------------
to_iter: <function to_iter at 0x104939f70>
to_dict: <function to_dict at 0x104977dc0>
typecheck: <function typecheck at 0x104d590d0>
literal_de_: <function literal_de_ at 0x104d49430>
from_iter: <function from_iter at 0x104935ca0>
from_dict: <function from_dict at 0x104923f70>


```

Reply:
Thanks, please remove it :pray:

Reply:
done, please review 🙇